### ISSUE TITLE
Meilisearch shows irrelevant results after updating to v0.27

### ISSUE THREAD

**Issue by KariUs123** [meilisearch/meilisearch#2420]
**Describe the bug**
Since I updated meilisearch to version 0.27, the search results have been worse. In previous version (v0.25) of meilisearch when you searched a keyword like `westpoint blender`, the results would return exact matches of westpoint blender and nothing more. In the current version(v0.27) when you search that keyword, the results ranks the exact matches of the keyword first followed by irrelevant search results like` westpoint air conditioner`,` westpoint cooker` and so on. Basically the search engine returns all documents with westpoint keyword.

Now the worst thing is when you switch keyword  from `westpoint blender` to `blender westpoint`, the search results will show westpoint blenders first followed by other documents with the keyword blender like `kenwood blender`, `evvoli blender` and so on.

If you search `westpoint blender` => results
```
[
    {
        "id": 1,
        "title": "Westpoint Blender ",
        "price": 200
    },
    {
        "id": 2,
        "title": "Westpoint air conditioner",
        "price": 520
    },
  {
        "id": 3,
        "title": "Westpoint  cooker",
        "price": 900
    }
]
```

**Expected behavior**
What I expect If I search `westpoint blender` => results
```
[
    {
        "id": 1,
        "title": "Westpoint Blender ",
        "price": 200
    }
    
]
```

**Meilisearch version:** [e.g. v0.27.1]

**Comment by MarinPostma** [comment]
That's definitely caused by the `words` criterion. pinging @gmourier

**Comment by gmourier** [comment]
What change has been made on this criterion to see such a change? Is it related to the first char being counted as two typos?

**Comment by curquiza** [comment]
Hello @KariUs123 

I've just tested to import these documents in v0.25.0

```json
[
    {
        "id": 1,
        "title": "Westpoint Blender ",
        "price": 200
    },
    {
        "id": 2,
        "title": "Westpoint air conditioner",
        "price": 520
    },
  {
        "id": 3,
        "title": "Westpoint  cooker",
        "price": 900
    }
]
```

and doing this search: 

```json
{
    "q": "westpoint blender"
}
```

and I get:

```json
{
    "hits": [
        {
            "id": 1,
            "title": "Westpoint Blender ",
            "price": 200
        },
        {
            "id": 2,
            "title": "Westpoint air conditioner",
            "price": 520
        },
        {
            "id": 3,
            "title": "Westpoint  cooker",
            "price": 900
        }
    ],
    "nbHits": 3,
    "exhaustiveNbHits": false,
    "query": "westpoint blender",
    "limit": 20,
    "offset": 0,
    "processingTimeMs": 2
}
```

I don't see any change between v0.25.0 and v0.27.0 regarding the relevancy with your example. Do you have another example to show the issue you have?

**Comment by curquiza** [comment]
Ok, if your version were not v0.27.0 but before v0.21.0, this is expected.
We indeed changed this behavior in v0.21.0. The `words` ranking rule is responsible for this. [More about it](https://docs.meilisearch.com/learn/core_concepts/relevancy.html#_1-words).
This is not something you cannot customize at the moment, but this is something we are considering. We have opened a discussion about this customization: https://github.com/meilisearch/product/discussions/255
You are more than welcome to give your needs, use casess and opinions on it 😄 

I close this issue since it's not a bug, but an expected behavior 😇 

pinging @gmourier (our PM) FYI

**Comment by curquiza** [comment]
Hello @KariUs123 and everyone here! 👋 

We are working on introducing the relevancy customization during the search with Meilisearch, and we just released a first prototype that allows you, for example, only to return documents containing ALL the query words! :fire:

It would be super helpful if you could let us know if this helps with your issue.
To know more about the prototype and how to use it, take a look at [the dedicated discussion](https://github.com/meilisearch/meilisearch/discussions/2639).

**Comment by curquiza** [comment]
Hello @KariUs123  👋

We have just released [v0.29.0rc1](https://github.com/meilisearch/meilisearch/releases/tag/v0.29.0rc1), which is a release candidate of v0.29.0 🔥

It introduces a new search parameter named `matchingStrategy` which allows you to change the default Meilisearch behavior about retrieving documents.

For now, it accepts 2 strategies:

- `last`: _the default strategy_. If no document contains all the given terms, the engine removes the last query term and so on.
- `all`: All the query terms must be present in a document to make it a search result.

How to use it:

```bash
curl \
  -X POST 'http://localhost:7700/indexes/movies/search' \
  -H 'Content-Type: application/json' \
  --data-binary '{ "q": "american ninja", "matchingStrategy": "all" }'
```

[Binaries](https://github.com/meilisearch/meilisearch/releases/tag/v0.29.0rc1) are attached to the release, or you can use the docker image:
```bash
docker run -it --rm \
    -p 7700:7700 \
    getmeili/meilisearch:v0.29.0rc1
```

Let us know about any bugs or feedback! 😄 It would be really helpful.

FYI, the official v0.29.0 release will be available on 3rd October.
